### DISCUSSION TITLE
Does Prisma support anything like Sequelize's Scopes?

### DISCUSSION THREAD

**Discussion by iwaffles** [prisma/prisma#10473] [Q&A] [answered]
Reference: https://sequelize.org/master/manual/scopes.html

The idea is that you can do something like `User.scope('deleted').findAll();` which would automatically setup the query to include:

```
where: {
    deleted: true
}
```

Or you could limit the fields you returned by saying `User.scope('public').findAll()`, which would only include selected attributes defined in the model. Something like:

```sql
select id, name from users limit 5;
```

instead of selecting all of the attributes.

**Comment by austincrim** [comment, marked as answer]
Hey @iwaffles 👋, thanks for the question.

Prisma doesn't offer a feature similar to scopes. If you find yourself needing to reuse common queries, you can always extract it to its own function or pull the query options into their own variable.

```ts
export async function getAdminUsers() {
  return await prisma.user.findMany({
    where: {
      role: 'ADMIN'
    }
  })
}
```

Do you have other use cases in mind that Prisma doesn't currently support?

**Comment by ortonomy** [comment]
Is there any plan to implement this @austincrim? it's a feature of the Eloquent ORM in Laravel, and the ActiveRecord ORM in rails, as well as sequelize. 

Moreover, you can make a subset of these scopes enabled by default. This is really useful for being able to say, ignore all deleted values and ONLY fetch them on special occasions, or vice versa... 

```
// pained laravel example, pseudo code

class SomeModel extends Model {
  // ....

  private static defaultScopes = ['onlyActive']

  public function scopeOnlyActive (Model model) {
    model.where({ deletedAt: false });
  }
}
```

usage:

```
$notDeleted = SomeModel::findAll();
$withDeleted = SomeModel::disableScopes(["onlyActive"]).findAll();
```

**Reply by austincrim** [reply]
That looks really interesting? Would you mind opening a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with more details?

This kind of behavior might be possible to implement on your own once we get further with [Prisma Client extensions](https://github.com/prisma/prisma/issues/7161), for what it's worth.

**Comment by Alsania14** [comment]
Hello i do something like this, we can make a function to define our scope in Prisma ORM.

```
export type CheckSelectKeys<T, U> = {
  [K in keyof T]: K extends keyof U ? T[K] : never;
};
```

```
export default function PrismaScope<R>() {
  abstract class Scope {
    public static generateScope = <T extends R>(arg: CheckSelectKeys<T, R>) =>
      arg;
  }
  return Scope;
}
```
and then we can extends function above like this

```
export default class SomethingModelScope extends PrismaScope<Prisma.SomethingModelSelect>() {
  static getPlain = this.generateScope({
    id: true,
    nama: true,
    tipe_parameter: true,
    preview_example: true,
  });
}
```

we can access our scope with a single line of code like this

```
const test = await this.client.SomethingModel.findUnique({
      where: {
        id: idParameter,
      },
      select: SomethingModelScope.getPlain,
    });
```

**Reply by normandanielmq** [reply]
Holy s#$# typescript took my joy for programming away.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
